### PR TITLE
Implement new host: "remote"

### PR DESCRIPTION
**Pull request by jugglinmike**
The "remote" host allows for executing code in any externally-hosted web browser via a remote Selenium/WebDriver server. This is tested entirely within TravisCI through a dedicated job that runs such a server locally.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -76,8 +82,18 @@ install: |
       unzip -d chrome chromedriver_linux64.zip
       PATH=$PATH:$(pwd)/chrome
       unset ESHOST_SKIP_CHROME;
+    elif [[ "$ESHOST_TARGET" == "remote" ]]; then
+      install_firefox;
+      wget http://selenium-release.storage.googleapis.com/3.4/selenium-server-standalone-3.4.0.jar
+      PATH=$PATH:$(pwd)/firefox java -jar selenium-server-standalone-3.4.0.jar &> selenium-server.log &
+
+      unset ESHOST_SKIP_REMOTE;
```
Comment (rwaldron) [inline .travis.yml]: This is rad

File: lib/agents/BrowserAgent.js [hunk 2]
```diff
@@ -4,14 +4,18 @@ const Server = require('../Server.js');
...
+    this.webHost = options.webHost || 'localhost';
+    this.webPort = options.webPort || 1337;
     this.id = agentId++;
   }
 
   initialize() {
     return Server.start(this.id).then(_ => {
-      this._url = 'http://localhost:1337/?clientId=' + this.id +
+      this._url = 'http://' + this.webHost + ':' + this.webPort +
+        '/?clientId=' + this.id +
```
Comment (rwaldron) [inline lib/agents/BrowserAgent.js]: nit: wdyt about template string cleanup on aisle 17 & 18? ;)

File: lib/agents/remote.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+const WebDriverAgent = require('../WebdriverAgent');
+const { WebDriver, Capabilities } = require('selenium-webdriver');
+const _http = require('selenium-webdriver/http');
+
+class RemoteAgent extends WebDriverAgent {
+  constructor(options) {
```
Comment (rwaldron) [inline lib/agents/remote.js]:
```js
constructor(options = {}) {
```

File: lib/agents/remote.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    super(options);
+
+    if (!options || typeof options.webdriverServer !== 'string') {
```
Comment (rwaldron) [inline lib/agents/remote.js]:
```diff
- !options
```

File: lib/agents/remote.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    }
+  }
+
+  _createDriver() {
+    const url = this.hostPath;
+    const client = Promise.resolve(new _http.HttpClient(this.webdriverServer));
+    const executor = new _http.Executor(client);
+    const expandedCaps = {
+      desired: this.capabilities,
+      required: this.capabilities
```
Comment (rwaldron) [inline lib/agents/remote.js]: Is this intentional?

File: lib/agents/remote.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        'required but was not specified.');
+    }
+    this.capabilities = Object.assign({}, options.capabilities);
+
+    if (typeof options.hostPath === 'string') {
+      throw new Error('RemoteAgent: \'hostPath\' specified.');
+    }
+  }
+
+  _createDriver() {
```
Comment (rwaldron) [inline lib/agents/remote.js]: nit: what does the `_` signify?

File: lib/agents/remote.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+const WebDriverAgent = require('../WebdriverAgent');
+const { WebDriver, Capabilities } = require('selenium-webdriver');
+const _http = require('selenium-webdriver/http');
```
Comment (rwaldron) [inline lib/agents/remote.js]: `_http` -> `swhttp`?

File: runtimes/browser.html [hunk 8]
```diff
@@ -18,7 +18,7 @@
 delete window.$;
 window[$.shortName] = $;
 
-$.socket = io("http://localhost:1337");
+$.socket = io(location.origin);
```
Comment (rwaldron) [inline runtimes/browser.html]: 👍

File: .travis.yml (hunk 1 above)
Comment (jugglinmike) [inline .travis.yml]: :D

File: lib/agents/BrowserAgent.js (hunk 2 above)
Comment (jugglinmike) [inline lib/agents/BrowserAgent.js]: Sure

File: lib/agents/remote.js (hunk 7 above)
Comment (jugglinmike) [inline lib/agents/remote.js]: Sure

File: lib/agents/remote.js (hunk 3 above)
Comment (jugglinmike) [inline lib/agents/remote.js]: Good idea

File: lib/agents/remote.js (hunk 6 above)
Comment (jugglinmike) [inline lib/agents/remote.js]: Usually it signifies that a method is not to be used by a library's consumers. That seems to be the case here, too.

In any event, it is not a name that I chose; I am overriding a method used by [the parent class](https://github.com/bterlson/eshost/blob/ad26299b4b797f70eed03d3d3b69b5e156359e1c/lib/WebdriverAgent.js).

File: lib/agents/remote.js (hunk 5 above)
Comment (jugglinmike) [inline lib/agents/remote.js]: Yes, it was necessary during my local testing.

**Comment by jugglinmike** [comment]
I think that's everything, @rwaldron.

File: lib/agents/remote.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+      throw new UnspecifiedOptionError('capabilities.platform');
+    }
+    if (typeof options.capabilities.version !== 'string') {
+      throw new UnspecifiedOptionError('capabilities.version');
+    }
+    this.capabilities = Object.assign({}, options.capabilities);
+
+    if (typeof options.hostPath === 'string') {
+      throw new UnspecifiedOptionError('hostPath');
+    }
```
Comment (rwaldron) [inline lib/agents/remote.js]: Beautiful!

File: lib/agents/remote.js (hunk 6 above)
Comment (rwaldron) [inline lib/agents/remote.js]: Gross, but fair enough :P

**Comment by rwaldron** [comment]
@bterlson I like this and my first review is satisfied, but it's a sizable new thing, so you should give it a once over as well.

**Comment by bterlson** [comment]
What do you have to do on the remote side to make this work?

Also I wonder if this is the right approach. I have a half-start on something like this which would talk to an eshost instance running on a remote host and expose all hosts that version of eshost is connected to. Imagine a build box that has binaries for all the console hosts that you can just add as a "remote" host in your local config and get access to all its hosts. Any thoughts on the relative merits of this PR vs,. an approach that enables remote console hosts as well?

**Comment by jugglinmike** [comment]
> What do you have to do on the remote side to make this work?

The necessary steps for setting up a "remote end" are demonstrated by this patch's modification to the `travis.yml` file. Essentially:

- Install one or more browser/driver pairs
- Execute the Selenium server `.jar` file

> Any thoughts on the relative merits of this PR vs,. an approach that enables
> remote console hosts as well?

This patch allows ESHost to interact with existing Selenium service providers like BrowserStack and Sauce Labs. While this has direct utility for me and my work, it's also relevant for the eshost project itself. That's because it will allow us to test integration with browsers like Edge and Safari, for which there does not (currently) exist any free continuous integration services.

I have another patch queued up that depends on this one which integrates a "free for Open Source projects" Sauce Labs account with this project. [You can view a recording of the tests passing in Safari here](https://saucelabs.com/beta/tests/464c283e8ef34dd9ac1c9b9cea0f6006/commands#28) (possibly one of the most dull videos ever recorded by Sauce Labs).

**Comment by bterlson** [comment]
@jugglinmike that is good rationale - using remote hosting services. Really cool. I guess this PR doesn't obviate my work for a "remote eshost". I'm happy considering them orthogonal.

+1 for this change!

**Comment by jugglinmike** [comment]
Great! I hadn't considered the "remote" architecture you described. If/when that is implemented, it will probably make sense to rename this host to something more specific like `remotewebdriver`.

**Pull request merged**